### PR TITLE
Tracing middleware

### PR DESCRIPTION
**Pull request by adam-fowler**
Middleware that adds span with basic tracing attributes. This does not add the V2 attributes as it seems V3 has ditched them. I haven't added the S3 and DynamoDB attributes yet as I haven't found a way to do this cleanly which I'm happy with.

This will need a rebuild of the services as I have added a new parameter to ServiceConfig which is the service name which differs slightly from the service identifier.

@slashmo

Ditched the previous PR as it was merging to the wrong branch

### CODE REVIEW

File: Sources/SotoCore/Middleware/Middleware/TracingMiddleware.swift [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    public init() {}
+
+    public func handle(
+        _ request: AWSHTTPRequest,
+        context: AWSMiddlewareContext,
+        next: (AWSHTTPRequest, AWSMiddlewareContext) async throws -> AWSHTTPResponse
+    ) async throws -> AWSHTTPResponse {
+        try await InstrumentationSystem.tracer.withSpan(
+            "\(context.serviceConfig.serviceName).\(context.operation)",
+            ofKind: .server
```
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/SotoCore/Middleware/Middleware/TracingMiddleware.swift:31]: Yep. Missed that

File: Sources/SotoCore/AWSServiceConfig.swift [hunk 2]
```diff
@@ -54,7 +56,7 @@ public final class AWSServiceConfig {
     ///   - region: Region of server you want to communicate with
     ///   - partition: Amazon endpoint partition. This is ignored if region is set. If no region is set then this is used along side partitionEndpoints to calculate endpoint
     ///   - amzTarget: "x-amz-target" header value
-    ///   - service: Name of service endpoint
```
Comment (0xTim) [inline Sources/SotoCore/AWSServiceConfig.swift:57]: You're missing the comment for `serviceName` here

**Review by 0xTim — APPROVED** [review]
Missing parameter in DocC comments otherwise LGTM

**Review by ktoso — APPROVED** [review]
Very nice usage of the tracing apis! LGTM :-)

**Pull request merged**